docs/man/memflush.1
.TH "MEMFLUSH" "1" "September 28, 2011" "0.53" "libmemcached"
.SH NAME
memflush \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Reset a server or list of servers
.SH SYNOPSIS
.sp
memflush [options]
.SH DESCRIPTION
.sp
\fBmemflush\fP  resets the contents of memcached(1) servers.
This means that all data in the specified servers will be deleted.
.SH OPTIONS
.sp
You can specify servers via the option:
.INDENT 0.0
.TP
.B \-\-servers
.UNINDENT
.sp
or via the environment variable:
.INDENT 0.0
.TP
.B \(gaMEMCACHED_SERVERS\(ga
.UNINDENT
.sp
For a full list of operations run the tool with option:
.INDENT 0.0
.TP
.B \-\-help
.UNINDENT
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH AUTHOR
.sp
Brian Aker, <\fI\%brian@tangent.org\fP>
.sp
Mark Atwood <\fI\%mark@fallenpegasus.com\fP>
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\"
.


ChangeLog
2024-09-02 [version 1.14.0]
## 🐣 New Features
* Add `mop upsert` method. #292
* Add multiple storage method.
  * `memcached_mset()`
  * `memcached_madd()`
  * `memcached_mreplace()`
  * `memcached_mprepend()`
  * `memcached_mpppend()`
  * `memcached_mcas()`
## 🐛 Bug Fixes
* Fix buffer size when calling `snprintf()` with leading whitespace and integer value. #304
* Use `__linux__` instead of `TARGET_OS_LINUX` to check if the current OS is Linux. #306
* Use different lock filenames for multiple proxy in a process
## 📝 Documentation
* Fix typo.
## ✅ Testing
* Cancel previous CI in progress when new commits are added. #296

2024-02-22 [version 1.13.3]
## 🔧 Enhancements
* Fix compile warnings on RHEL8 and MacOS.
* Do not use `#pragma once`.
* Select smallest host as owner when ketama hash collided. #237
* Add parameter validation in `arcus_proxy_connect()`.
* Move io reset from where calls `memcached_vdo()` to where fails inside of `memcached_vdo()`. #250
* Use `<latest-release>-unknown` version when git describe fails.
* Add a limit of count about bop mget/smget commands.
* Import `scanner.l` and `parser.yy` from libmemcached-0.53.
* Refactor ketama weighted related implementation. #263 #273 #274 #275 #276 #277
* Fix package name and info to 'arcus-c-client'.
## 🐛 Bug Fixes
* Call `arcus_server_check_for_update()` in `memcached_exist_by_key()`. #241
* Check `mc->connect_timeout` instead of `mc->poll_timeout` in `connect_poll()`. #247
* Remove a duplicate init macro in configure.ac. #267
* Return `hashkit_get_distribution_function()` instead of `hashkit_get_function()` in `memcached_behavior_get_distribution_hash()`.
## 📝 Documentation
* Add unit test FAQ.
* Fix inconsistencies between code and documentation.
## ✅ Testing
* Use Github Actions instead of Travis for CI test.

2022-02-22 [version 1.13.2]
        * [FEATURE] supported the 3 replicas in replication.
        * [FIX] return switchover error code when coll get with delete.
        * [FIX] fixed wrong aggregate pipe return code in case of failure.
        * [FIX] hold the pool lock when update the cache list of master mc.
        * [FIX] fixed the cache list update of member mc in proxy mode.
        * [ENHANCE] enhanced the cachelist update logic with serverinfo.
        * [ENHANCE] Not to close and reopen pool connections
        * [ENHANCE] removed the dual pool repopulation when start at first.
        * [ENHANCE] optimized the slave comparision in memcached_rgroup_update().
        * [ENHANCE] optimized the release processing in memcached_pool.
        * [ENHANCE] ensure more concurrency in memcached pool.
        * [CLEANUP] fixed the compile warnings.
        * [CLEANUP] refactored serveral parts of code.

2021-07-06 [version 1.13.1]
        * [FEATURE] enable zookeeper multi thread mode.
        * [FIX] if MEMCACHED_ERRNO entered in memcached_set_error(),
                return instead of assert.

2021-01-18 [version 1.13.0]
        * [FEATURE] supported mgets command.
        * [FIX] check space_sepatated_keys immediately after memcached_connect().
        * [FIX] check mget command support immediately after memcached_connect().
        * [INTERNAL] return MEMCACHED_CONNECTION_FAILURE in memcached_fetch_result().

2020-12-09 [version 1.12.0]
        * [FEATURE] set the max key size to 4000.
        * [DOC] changed max element bytes: 4KB => 16KB.
                (committed at 20200317 and included in 1.10.4)
        * [DOC] fixed the smget result API and structure names.

2020-12-07 [version 1.11.0]
        * [FEATURE] supported space-separated keys.
        * [FEATURE] supported mget command, efficient for getting multiple items.
        * [FEATURE] requested version operation when connect to server.
        * [FEATURE] increased max key length: 250 => 4K.
        * [ENHANCE] do memcached_quit_server() if poll timeout occurs in io_flush.
        * [ENHANCE] increased cluster max size for multi key op: 200 => 1000.
        * [IMPROVE] reduced string copy when convert eflag/update_filter to string.
        * [FIX] set purge recursion flag to false before return.
        * [FIX] handled the host failures in the multi key operation.
        * [FIX] unset noreply before get operation in noreply_test().
        * [FIX] fixed unallocated memory access in mkey fetch.
        * [OTHERS] Refactored some code for readabiblity.

2020-10-22 [version 1.10.5]
        * [FEATURE] bop incr/decr with initial option.
        * [FIX] registered "NULL" if hostname lookup fails in do_add_client_info().
        * [FIX] added the TYPE_MISMATCH handling in kv response
        * [DOC] changed the description of the max item size
        * [DOC] refactored the document format and some content.

2020-08-04 [version 1.10.4]
        * [FEATURE] Added memcached_detail_error_message() API.
        * [FEATURE] Immediate reconnect according to error responses.
        * [FEATURE] Changed max element bytes. 4KB => 16KB.
        * [FEATURE] Returned E2BIG if CLIENT_ERROR object too large.
        * [FEATURE] Changed MEMCACHED_SERVER_FAILURE_RETRY_TIMEOUT. 2sec => 1sec.
        * [ENHANCE] Enlarged the contents of error messages in detail.
        * [ENHANCE] Stored detail error message in error_messages field.
        * [ENHANCE] Added arcus ZK manager for processing ZK events.
        * [FIX] pipe operation is stopped if switchover is done.
        * [FIX] Fixed version check condition.
        * [FIX] Fixed several bugs in handling response strings.
        * [FIX] Added the key test where needed.
        * [FIX] Do client_info and watcher registration only when needed.
        * [FIX] Prevent master_mc from being freed while using pool.
        * [FIX] Changed to create version using m4 file.
        * [FIX] Fixed simple null pointer dereferencing.
        * [DOC] Added multi eflag filter description.
        * [Test] Added test case for PIPE_ERROR bad error.
        * [OTHERS] Refactored the response module for handling responses.

clients/memflush.cc
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary:
 *
 */
#include "config.h"

#include <cstdio>
#include <cstring>
#include <getopt.h>
#include <iostream>
#include <unistd.h>

#include <libmemcached/memcached.h>
#include "client_options.h"
#include "utilities.h"

static int opt_binary= 0;
static int opt_verbose= 0;
static time_t opt_expire= 0;
static char *opt_servers= NULL;
static char *opt_username;
static char *opt_passwd;

#define PROGRAM_NAME "memflush"
#define PROGRAM_DESCRIPTION "Erase all data in a server of memcached servers."

/* Prototypes */
void options_parse(int argc, char *argv[]);

int main(int argc, char *argv[])
{
  options_parse(argc, argv);

  if (opt_servers == NULL)
  {
    char *temp;

    if ((temp= getenv("MEMCACHED_SERVERS")))
    {
      opt_servers= strdup(temp);
    }
    else
    {
      std::cerr << "No Servers provided" << std::endl;
      exit(EXIT_FAILURE);
    }
  }

  memcached_st *memc= memcached_create(NULL);

  memcached_server_st *servers= memcached_servers_parse(opt_servers);
  memcached_server_push(memc, servers);
  memcached_server_list_free(servers);
  memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL,
                         (uint64_t) opt_binary);

  if (opt_username and LIBMEMCACHED_WITH_SASL_SUPPORT == 0)
  {
    memcached_free(memc);
    std::cerr << "--username was supplied, but binary was not built with SASL support." << std::endl;
    return EXIT_FAILURE;
  }

  if (opt_username)
  {
    memcached_return_t ret;
    if (memcached_failed(ret= memcached_set_sasl_auth_data(memc, opt_username, opt_passwd)))
    {
      std::cerr << memcached_last_error_message(memc) << std::endl;
      memcached_free(memc);
      return EXIT_FAILURE;
    }
  }

  memcached_return_t rc = memcached_flush(memc, opt_expire);
  if (rc != MEMCACHED_SUCCESS)
  {
    std::cerr << memcached_last_error_message(memc) << std::endl;
  }

  memcached_free(memc);

  free(opt_servers);

  return EXIT_SUCCESS;
}


void options_parse(int argc, char *argv[])
{
  static struct option long_options[]=
  {
    {(OPTIONSTRING)"version", no_argument, NULL, OPT_VERSION},
    {(OPTIONSTRING)"help", no_argument, NULL, OPT_HELP},
    {(OPTIONSTRING)"quiet", no_argument, NULL, OPT_QUIET},
    {(OPTIONSTRING)"verbose", no_argument, &opt_verbose, OPT_VERBOSE},
    {(OPTIONSTRING)"debug", no_argument, &opt_verbose, OPT_DEBUG},
    {(OPTIONSTRING)"servers", required_argument, NULL, OPT_SERVERS},
    {(OPTIONSTRING)"expire", required_argument, NULL, OPT_EXPIRE},
    {(OPTIONSTRING)"binary", no_argument, NULL, OPT_BINARY},
    {(OPTIONSTRING)"username", required_argument, NULL, OPT_USERNAME},
    {(OPTIONSTRING)"password", required_argument, NULL, OPT_PASSWD},
    {0, 0, 0, 0},
  };

  bool opt_version= false;
  bool opt_help= false;
  int option_index= 0;
  while (1)
  {
    int option_rv= getopt_long(argc, argv, "Vhvds:", long_options, &option_index);
    if (option_rv == -1) break;
    switch (option_rv)
    {
    case 0:
      break;

    case OPT_BINARY:
      opt_binary= true;
      break;

    case OPT_VERBOSE: /* --verbose or -v */
      opt_verbose= OPT_VERBOSE;
      break;

    case OPT_DEBUG: /* --debug or -d */
      opt_verbose= OPT_DEBUG;
      break;

    case OPT_VERSION: /* --version or -V */
      opt_version= true;
      break;

    case OPT_HELP: /* --help or -h */
      opt_help= true;
      break;

    case OPT_SERVERS: /* --servers or -s */
      opt_servers= strdup(optarg);
      break;

    case OPT_EXPIRE: /* --expire */
      opt_expire= (time_t)strtoll(optarg, (char **)NULL, 10);
      break;

    case OPT_USERNAME:
      opt_username= optarg;
      break;

    case OPT_PASSWD:
      opt_passwd= optarg;
      break;

    case OPT_QUIET:
      close_stdio();
      break;

    case '?':
      /* getopt_long already printed an error message. */
      exit(EXIT_FAILURE);

    default:
      abort();
    }
  }

  if (opt_version)
  {
    version_command(PROGRAM_NAME);
    exit(EXIT_SUCCESS);
  }

  if (opt_help)
  {
    help_command(PROGRAM_NAME, PROGRAM_DESCRIPTION, long_options, NULL);
    exit(EXIT_SUCCESS);
  }
}
